Forward subscription cancellation from client to server

alerosmile opened the pull request:
This PR forwards the cancellation events from response `Uni` and `Multi` to the server by cancelling the gRPC stream.

- Fixes: #41776 
- Fixes: #46986.

alerosmile committed:
Forward subscription cancellation from client to server.

alesj commented:
Plenty of CI failures …

On Mon, 28 Apr 2025 at 10:31, Georgios Andrianakis ***@***.***>
wrote:

> @geoand <https://github.com/geoand> requested your review on: #47555
> <https://github.com/quarkusio/quarkus/pull/47555> Forward subscription
> cancellation from client to server.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/quarkusio/quarkus/pull/47555#event-17429857653>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AACRA6F2GWJYPJQKP3UXJTD23XRP7AVCNFSM6AAAAAB34J24R6VHI2DSMVQWIX3LMV45UABCJFZXG5LFIV3GK3TUJZXXI2LGNFRWC5DJN5XDWMJXGQZDSOBVG43DKMY>
> .
> You are receiving this because your review was requested.Message ID:
> ***@***.***>
>

cescoffier reviewed (dismissed):
LGTM, let's wait for @alesj and @jponge opinions.

alerosmile commented:
I've set this to draft because I'll provide some integration tests for this.

alerosmile commented:
@cescoffier please see https://github.com/quarkusio/quarkus/issues/41776#issuecomment-2831216389. What are the classes `ManyToManyObserver` and `ManyToOneObserver` used for?

alerosmile committed:
Add integration tests (some tests are disabled).

O2N and Vert.x tests are disable because they fail (see https://github.com/quarkusio/quarkus/pull/47555#issuecomment-2843058567 for details)

jponge commented on extensions/grpc/stubs/src/main/java/io/quarkus/grpc/stubs/MultiStreamObserver.java:
```diff
@@ -1,28 +1,62 @@
+    // If this is set, the Multi was terminated by completion or failure, not by cancellation.
+    private final AtomicBoolean terminated = new AtomicBoolean();
+
+    private AtomicReference<ClientCallStreamObserver<I>> requestStreamObserver = new AtomicReference<>();
```
You might be able to condense the 2 atomic variables into one (`requestStreamObserver`, when `null` it means terminated)

alerosmile replied:
Is it guaranteed that `beforeStart` is called in every case? If not, the emitter will never be terminated on error (after the change, `emitter.fail()` is only called if `requestStreamObserver` is not `null`).

alerosmile commented:
Added some integration test. The tests in which Vert.x is used on the server side are disabled because they fail (see https://github.com/quarkusio/quarkus/pull/47555#issuecomment-2843058567).

alerosmile commented:
@geoand no, unfortunately not yet. I haven't found the time to comment on https://github.com/eclipse-vertx/vert.x/issues/5562. This has to be solved otherwise this PR doesn't make any sense (see my previous comments here). Is anybody from the Quarkus core team here to help on the Vert.x issue?

geoand commented:
Maybe @jponge or @alesj can help

cescoffier commented:
@jponge FYI.

jbgomond commented:
Hello, we are facing this issue too. Do you know when it's due to be merged ? Thanks :)

gsmet commented:
Not sure if that’s all that’s needed but we updated to Vert.x 4.5.25 in 3.31.4.

alerosmile committed:
Re-enabling O2N and Vert.x tests (bug in Vert.x has been fixed)

alerosmile commented:
Unfortunately, the fix does not resolve the issue (see https://github.com/eclipse-vertx/vert.x/issues/5562#issuecomment-3959685420).

anavarr commented:
I had a look at it this afternoon and got an interesting lead [here](https://github.com/quarkusio/quarkus/issues/46986#issuecomment-4245522743).

jmini commented:
`4.5.27` hasn't been released yet… https://github.com/eclipse-vertx/vertx-grpc/tags

jponge commented:
@jmini see https://github.com/quarkusio/quarkus/pull/53975

jponge commented:
@anavarr Could you please have a look? 🙏

anavarr reviewed (approved):
significant overlap with https://github.com/quarkusio/quarkus/pull/53313. Simpler to merge this one first and 53313 after.

anavarr commented:
hey @alerosmile. Your PR looks good.
It has significant overlap with https://github.com/quarkusio/quarkus/pull/53313 concerning `Multi` handling in `ServerCalls` and `ClientCalls`. Yours addresses cancellation, mine flow control. I feel like yours has a narrower scope, landing yours then mine seems simpler than the opposite. I think we can merge it @jponge @cescoffier.

jponge commented:
Thanks @anavarr, I've launched CI to see how it goes.

/cc @geoand

geoand commented:
Nice work!

The CI status is odd...

anavarr commented:
@jponge can you restart the CI ?

geoand commented:
I restarted all jobs

jponge commented:
Sorry guys, I was busy on another PR

jponge reviewed (approved):
We will then let @anavarr update his own PR to adresse flow-control issues that we've identified elsewhere. Thanks @alerosmile
